torch/distributed/_tensor/api.py

```diff
@@ -252,6 +261,91 @@ def __new__(
+                full_shard_size.append(full_chunk_size)
+
+                # Currently we assume the shard id is to equal to the global rank.
+                # Note that this might not be true for 2D (strided sharding + global_rank)
+                # and 3D (strided sharding + 2D local rank).
+                cur_chunk = mesh.get_rank()
+
+                # We cannot directly use the shape of local_tensor as cur_unpadded_shard_size, since we need to verify
+                # whether the shape of local_tensor given by user is legit.
+                # We need to derive the shape of the current shard without padding based on the global shape and the shard id.
+
+                # If the index of cur chunk is smaller than num_full_chunks,
+                # this means cur_chunk would be a full chunk on the given tensor dimension.
+                if cur_chunk < num_full_chunks:
+                    cur_unpadded_shard_size.append(full_chunk_size)
+                # If the index of cur_chunk is num_full_chunks and the tail_chunk_size is not 0,
+                # this means the cur_chunk is the non-empty tail chunk.
+                # There should be only 1 non-empty tail chunk.
+                # For example, shard [1, 1, 1, 1, 1] to 4 chunks, we would have [1, 1], [1, 1], [1].
+                # The third shard is a non-empty tail chunk and the last shard is an empty chunk.
+                elif cur_chunk == num_full_chunks and tail_chunk_size != 0:
+                    cur_unpadded_shard_size.append(tail_chunk_size)
+                # Otherwise, the cur_chunk is an empty chunk on the tensor_dim. There could be more than 1 empty chunks.
+                # For example, chunk a tensor([1, 1]) into 4 chunks, the last two chunks would be empty.
+                else:
+                    cur_unpadded_shard_size.append(0)
+
+            # 3. add a check to see whether the given local tensor is legit.
+            if list(local_tensor.shape) != cur_unpadded_shard_size:
+                raise RuntimeError(
+                    f"The given local tensor shape {local_tensor.shape} does not"
+                    f"match the expected local tesnsor shape {cur_unpadded_shard_size}!"
+                )
+
+            # 4. short-circuit return if no padding is needed for the given shard.
+            if full_shard_size == cur_unpadded_shard_size:
+                # make `unpadded_local_shard_size` a cached property in DTensorSpec
+                # so we don't need to re-calculate this during backward of _ToTorchTensor.
+                spec.unpadded_local_shard_size = None
+                return local_tensor, spec
+
+            # 5. create a padded local tensor and copy local tensor into padded local tensor
+            spec.unpadded_local_shard_size = (
+                cur_unpadded_shard_size  # make this a cached property in DTensorSpec
+            )
+            padded_local_tensor = local_tensor.new_zeros(full_shard_size)
+            slices = [
```

Comment:
why we still need to calculate the slice here? IMO cur_unpadded_shard_size should just be directly used as slice? if not we should just return the `slices` directly?

Reply:
This is just to generate the slice index for us. 
If we know the tensor dim, we can simply do something like, tensor[:2, :3, :4], but here we are dynamically generating the slices based on cur_unpadded_shard_size. 

If we need to add a property to DTensorSpec, cur_unpadded_shard_size would be more readable then this slice object.